kvbc/test/categorization/block_merkle_category_unit_test.cpp

```diff
@@ -466,6 +466,131 @@ TEST_F(block_merkle_category, prune_many_nodes) {
+  }
+  ASSERT_EQ(5, cat.getLatestTreeVersion());
+  ASSERT_EQ(5, cat.getLatestVersion(key1)->version);
+  auto expected5 = MerkleValue{{5, val1}};
+  ASSERT_EQ(expected5, asMerkle(*cat.getLatest(key1)));
+
+  // 5 blocks = 5 leaves
+  auto all_leaves_5 = getAllLeaves(db);
+  ASSERT_EQ(5, all_leaves_5.size());
+  // 5 blocks = 5 stale nodes
+  auto all_stale_5 = getAllStale(db);
+  ASSERT_EQ(5, all_stale_5.size());
+
+  // Ensure the stale node keys are actually just tree versions 1-5
+  auto count = 1;
+  for (auto &kv : all_stale_5) {
+    auto ver = TreeVersion{};
+    deserialize(kv.first, ver);
+    ASSERT_EQ(ver.version, count);
+    count++;
+  }
+
+  auto all_internal_5 = getAllInternalNodes(db);
+
+  auto batch = db->getBatch();
+  cat.deleteLastReachableBlock(5, out[4], batch);
+  db->write(std::move(batch));
+
+  auto all_leaves_4 = getAllLeaves(db);
+  ASSERT_EQ(4, all_leaves_4.size());
+  auto all_stale_4 = getAllStale(db);
+  ASSERT_EQ(4, all_stale_4.size());
+
+  // Ensure the stale node keys are actually just tree versions 1-4
+  count = 1;
+  for (auto &kv : all_stale_4) {
+    auto ver = TreeVersion{};
+    deserialize(kv.first, ver);
+    ASSERT_EQ(ver.version, count);
+    count++;
+  }
+
+  // There are fewer internal nodes now
+  auto all_internal_4 = getAllInternalNodes(db);
+  ASSERT_LT(all_internal_4.size(), all_internal_5.size());
+
+  ASSERT_EQ(4, cat.getLatestTreeVersion());
+  ASSERT_EQ(4, cat.getLatestVersion(key1)->version);
+  auto expected4 = MerkleValue{{4, val1}};
+  ASSERT_EQ(expected4, asMerkle(*cat.getLatest(key1)));
+
+  // Adding back block 5 causes the merkle column familes to be exactly the same as before the removal of block 5
+  auto update = BlockMerkleInput{{{key1, val1}}};
+  auto new_out = add(5, std::move(update));
+
+  ASSERT_EQ(all_leaves_5, getAllLeaves(db));
+  ASSERT_EQ(all_stale_5, getAllStale(db));
+  ASSERT_EQ(all_internal_5, getAllInternalNodes(db));
+  ASSERT_EQ(5, cat.getLatestTreeVersion());
+  ASSERT_EQ(5, cat.getLatestVersion(key1)->version);
+  ASSERT_EQ(expected5, asMerkle(*cat.getLatest(key1)));
+
+  // Deleting again generates the same column families as the prior delete
+  {
+    auto batch = db->getBatch();
+    cat.deleteLastReachableBlock(5, new_out, batch);
+    db->write(std::move(batch));
+    ASSERT_EQ(all_leaves_4, getAllLeaves(db));
+    ASSERT_EQ(all_stale_4, getAllStale(db));
+    ASSERT_EQ(all_internal_4, getAllInternalNodes(db));
+    ASSERT_EQ(4, cat.getLatestTreeVersion());
+    ASSERT_EQ(4, cat.getLatestVersion(key1)->version);
+    ASSERT_EQ(expected4, asMerkle(*cat.getLatest(key1)));
+  }
+
+  // Deleting all blocks does the right thing
+  {
+    auto batch = db->getBatch();
+    for (auto i = 4u; i > 0; i--) {
+      cat.deleteLastReachableBlock(i, out[i - 1], batch);
+    }
+    db->write(std::move(batch));
+    ASSERT_EQ(0, getAllLeaves(db).size());
+    // There's always a key pointing to the latest root
+    ASSERT_EQ(1, getAllInternalNodes(db).size());
```

Comment:
Nit: Based on our offline discussion, we might want to be stricter and not allow deletion of the last reachable block that is also genesis and, hence, relax these checks here.

Reply:
Fixed.